A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,398 @@
+           return subchannel as the pick;
+         }
+         equEquivalentAddresses = equivalent addresses from entry;
+         newAddresses = subset of equEquivalentAddresses not present in processedSet;
+         add newAddresses to equivalentAddresses at the end;
+       }
+     }
+     /* no suitable entry found in overrideHostMap */
+     if overrideHost address is not in validEndpoints map {
+       return subchannel from delegated child policy;
+     }
+     endpointEntry = validEndpoints entry for overrideHost;
+     create a subchannel with overrideHost and equivalentAddresses from endpointEntry;
+     if and when subchannel goes READY then {
+       add it to overrideHostMap and return as the pick;
+     } else {
+       /* subchannel is in TRANSIENT_FAILURE */
+       return subchannel from delegated child policy;
+     }
+   }
+```
+
+In the above logic we prefer a subchannel in the `READY` state for a different
+equivalent address instead of waiting for the subchannel for the original
+`overrideHost` address to become `READY` (from one of `IDLE`, `CONNECTING`
+or `TRANSIENT_FAILURE` states) because we assume that the equivalent address
+is pointing to the same host thereby maintaining session affinity with the
+pick.
+
+If we do not find an entry in `overrideHostMap` for the original or an
+equivalent address then we create a new subchannel and wait for that subchannel
+to become `READY` before returning that subchannel as the pick. By waiting,
+we are accepting the possibility of delays in favor of not immediately failing
+over to another host because delivering to the intended host is necessary
+for maintaining session affinity. If the new subchannel goes into
+`TRANSIENT_FAILURE`, we just delegate to the child policy.
+
+Note that we create the `“override-host-experimental”` policy as the child of
+`“xds_cluster_impl”` only when the `override_host_status` config is present
+which is passed through
+[`common_lb_config.override_host_status`][or-host-status]. The
+`“cluster_resolver_experimental”` policy will receive this value in its CDS
+update and pass this config via its child (`“priority_experimental”`) to the
+`“xds_cluster_impl”` policy. Only when `override_host_status` is set it will
+create the `“override-host-experimental”` policy as its child and make one of
+the existing policies (`"wrr-locality-experimental"`, `"weighted_target"`
+or `"ring_hash_experimental"`) as the child policy of
+`“override-host-experimental”` as shown in the following diagram.
+
+![LB Policy Hierarchy Diagram](A55_graphics/lb-hierarchy.png)
+
+To propagate [`common_lb_config.override_host_status`][or-host-status] the
+following changes are required:
+
+* the [`XdsClient`][grpc-client-arch] to parse
+[`common_lb_config.override_host_status`][or-host-status] and copy the value
+in its CDS update to the watchers.
+
+* `cds_experimental` policy (which receives the CDS update) copies the value
+into `ClusterResolverConfig` for its child policy
```

Comment:
`ClusterResolverConfig` looks like a Java-specific object.  I don't think it needs to be mentioned here; the important point here is just that the CDS policy will use the field in the `XdsClient` CDS update to populate the relevant field in the `xds_cluster_resolver` LB policy config.

Reply:
fixed